File: apps/docsign-tauri/src/updater.ts
/**
 * Auto-update module for the GetSignatures desktop application.
 *
 * This module provides TypeScript bindings for the native auto-update commands.
 * Designed with geriatric UX in mind:
 * - Clear, reassuring notifications
 * - Simple one-click update process
 * - Progress indication during download
 * - Safety messages about document preservation
 */

import { invoke } from "@tauri-apps/api/core";

/**
 * Information about an available update.
 */
export interface UpdateInfo {
  /** The new version number (e.g., "1.2.0") */
  version: string;
  /** Release notes describing what's new (optional) */
  notes?: string;
  /** When the update was released (optional, formatted date string) */
  date?: string;
}

/**
 * Error thrown when an update operation fails.
 */
export class UpdateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UpdateError";
  }
}

/**
 * Checks if a newer version of the application is available.
 *
 * This contacts the update server to check for newer versions.
 * Safe to call frequently - it's a lightweight check.
 *
 * @returns UpdateInfo if an update is available, null if already up to date.
 * @throws UpdateError if the check fails (e.g., no internet connection).
 *
 * @example
 * ```typescript
 * const update = await checkForUpdates();
 * if (update) {
 *   console.log(`Update available: ${update.version}`);
 *   console.log(`Release notes: ${update.notes}`);
 * } else {
 *   console.log('You have the latest version!');
 * }
 * ```
 */
export async function checkForUpdates(): Promise<UpdateInfo | null> {
  try {
    return await invoke<UpdateInfo | null>("check_for_updates");
  } catch (error) {
    throw new UpdateError(formatUpdateError(error));
  }
}

/**
 * Downloads and installs an available update.
 *
 * This will:
 * 1. Download the update package (may take a few minutes)
 * 2. Verify the download is safe and complete
 * 3. Install the update
 * 4. Restart the application automatically
 *
 * IMPORTANT: Make sure to save any work before calling this function.
 * The application will restart after the update is installed.
 *
 * @throws UpdateError if the update fails.
 *
 * @example
 * ```typescript
 * try {
 *   // Show progress indicator to user
 *   showProgress('Downloading update...');
 *
 *   await installUpdate();
 *   // Note: This line won't be reached - app restarts after update
 * } catch (error) {
 *   showError('Update failed. Please try again later.');
 * }
 * ```
 */
export async function installUpdate(): Promise<void> {
  try {
    await invoke<void>("install_update");
  } catch (error) {
    throw new UpdateError(formatUpdateError(error));
  }
}

/**
 * Gets the currently installed version of the application.
 *
 * Useful for displaying "Current version: X.Y.Z" in settings or about dialogs.
 *
 * @returns The current version string (e.g., "0.1.0").
 *
 * @example
 * ```typescript
 * const version = await getCurrentVersion();
 * console.log(`GetSignatures version ${version}`);
 * ```
 */
export async function getCurrentVersion(): Promise<string> {
  try {
    return await invoke<string>("get_current_version");
  } catch {
    return "unknown";
  }
}

/**
 * Checks if running in a Tauri environment where updates are supported.
 *
 * Updates are only available in the desktop app, not the web version.
 *
 * @returns true if updates are supported, false otherwise.
 */
export function isUpdateSupported(): boolean {
  return typeof window !== "undefined" && "__TAURI__" in window;
}

// ============================================================================
// Geriatric UX Helper Functions
// ============================================================================

/**
 * User-friendly message strings for update UI.
 * Designed for clarity and reassurance.
 */
export const UpdateMessages = {
  /** Shown when an update is available */
  updateAvailable: (version: string) =>
    `A new version (${version}) is available. Would you like to update now?`,

  /** Shown during download */
  downloading: "Downloading update... This may take a few minutes.",

  /** Shown when download is complete */
  installing: "Installing update... The app will restart shortly.",

  /** Reassurance about document safety */
  documentsSafe:
    "Your documents are safe. They will still be here after the update.",

  /** When already on latest version */
  upToDate: "You have the latest version. No update needed.",

  /** When update check fails */
  checkFailed:
    "Could not check for updates. Please check your internet connection.",

  /** When update installation fails */
  installFailed: "Update failed. Please try again later.",

  /** Prompt before update */
  updatePrompt:
    "A new version is available with improvements and fixes. Updating takes just a few minutes.",

  /** Button text for update */
  updateButton: "Update Now",

  /** Button text for later */
  laterButton: "Remind Me Later",

  /** Button text for skip */
  skipButton: "Skip This Version",
} as const;

/**
 * Formats release notes for display to elderly users.
 *
 * Simplifies technical release notes into user-friendly language.
 *
 * @param notes - Raw release notes from the update server
 * @returns Simplified, user-friendly summary
 */
export function formatReleaseNotes(notes: string | undefined): string {
  if (!notes) {
    return "This update includes improvements and fixes.";
  }

  // Keep it simple - just show the notes, but clean them up
  return notes
    .replace(/#+\s*/g, "") // Remove markdown headers
    .replace(/\*\*/g, "") // Remove bold markers
    .replace(/`[^`]*`/g, "") // Remove code blocks
    .replace(/\n{3,}/g, "\n\n") // Collapse multiple newlines
    .trim();
}

/**
 * Checks for updates and returns a user-friendly status.
 *
 * This is a convenience function that wraps checkForUpdates with
 * geriatric-friendly error handling and messaging.
 *
 * @returns An object with status information suitable for UI display
 */
export async function checkUpdateStatus(): Promise<{
  hasUpdate: boolean;
  message: string;
  updateInfo?: UpdateInfo;
}> {
  if (!isUpdateSupported()) {
    return {
      hasUpdate: false,
      message: "Updates are only available in the desktop app.",
    };
  }

  try {
    const update = await checkForUpdates();

    if (update) {
      return {
        hasUpdate: true,
        message: UpdateMessages.updateAvailable(update.version),
        updateInfo: update,
      };
    }

    return {
      hasUpdate: false,
      message: UpdateMessages.upToDate,
    };
  } catch {
    return {
      hasUpdate: false,
      message: UpdateMessages.checkFailed,
    };
  }
}

/**
 * Formats an error from the update backend into a user-friendly message.
 */
function formatUpdateError(error: unknown): string {
  if (typeof error === "string") {
    return error;
  }
  if (error instanceof Error) {
    return error.message;
  }
  if (typeof error === "object" && error !== null && "message" in error) {
    return String((error as { message: unknown }).message);
  }
  return "An unexpected error occurred during the update";
}

File: apps/docsign-tauri/src-tauri/src/commands/updater.rs
//! Auto-update functionality for the docsign desktop application.
//!
//! This module provides Tauri commands for checking and installing updates.
//! Designed with geriatric UX in mind:
//! - Clear, non-technical update notifications
//! - Simple one-click update process
//! - Reassurance that documents are safe during updates

use serde::Serialize;
use tauri::AppHandle;
use tauri_plugin_updater::UpdaterExt;

/// Information about an available update.
///
/// Serialized to JavaScript with user-friendly field names.
#[derive(Debug, Clone, Serialize)]
pub struct UpdateInfo {
    /// The new version number (e.g., "1.2.0")
    pub version: String,
    /// Release notes explaining what's new
    pub notes: Option<String>,
    /// When the update was released
    pub date: Option<String>,
}

/// Checks if an update is available for the application.
///
/// This command contacts the update server to check for newer versions.
/// Returns update information if available, or null if already up to date.
///
/// # Returns
/// - `Ok(Some(UpdateInfo))` - An update is available
/// - `Ok(None)` - Already running the latest version
/// - `Err(String)` - Failed to check for updates (network error, etc.)
#[tauri::command]
pub async fn check_for_updates(app: AppHandle) -> Result<Option<UpdateInfo>, String> {
    let updater = app
        .updater()
        .map_err(|e| format!("Update check failed: {}", e))?;

    match updater.check().await {
        Ok(Some(update)) => {
            // Parse the date from the update if available
            // The date is an OffsetDateTime, we just use its Display implementation
            let date = update.date.map(|d| d.to_string());

            Ok(Some(UpdateInfo {
                version: update.version.clone(),
                notes: update.body.clone(),
                date,
            }))
        }
        Ok(None) => {
            // No update available - already on latest version
            Ok(None)
        }
        Err(e) => {
            // Network or server error
            Err(format!(
                "Could not check for updates. Please check your internet connection. ({})",
                e
            ))
        }
    }
}

/// Downloads and installs an available update.
///
/// This will:
/// 1. Download the update package
/// 2. Verify its integrity
/// 3. Install the update
/// 4. Restart the application
///
/// # Important
/// Documents are automatically saved before the update process begins.
/// The application will restart automatically after the update is installed.
///
/// # Returns
/// - `Ok(())` - Update was downloaded and will be installed on restart
/// - `Err(String)` - Update failed with a user-friendly error message
#[tauri::command]
pub async fn install_update(app: AppHandle) -> Result<(), String> {
    let updater = app
        .updater()
        .map_err(|e| format!("Could not start update: {}", e))?;

    // Check for the update first
    let update = updater
        .check()
        .await
        .map_err(|e| format!("Could not check for updates: {}", e))?
        .ok_or_else(|| "No update available".to_string())?;

    // Download and install the update
    // The closure receives download progress events
    update
        .download_and_install(
            |_chunk_length, _content_length| {
                // Progress callback - could be used for UI progress bar
                // For now, the frontend handles progress indication separately
            },
            || {
                // Download complete callback
                // The update is ready to be applied
            },
        )
        .await
        .map_err(|e| format!("Update download failed. Please try again later. ({})", e))?;

    // Request app restart to apply the update
    // Note: On some platforms this happens automatically
    app.restart();
}

/// Gets the current application version.
///
/// Useful for displaying "Current version: X.Y.Z" in the UI.
#[tauri::command]
pub fn get_current_version(app: AppHandle) -> String {
    app.config()
        .version
        .clone()
        .unwrap_or_else(|| "unknown".to_string())
}
